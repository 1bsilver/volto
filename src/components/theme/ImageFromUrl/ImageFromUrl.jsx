import React from 'react';
import PropTypes from 'prop-types';
import {
  isInternalURL,
  flattenToAppURL,
  getSrcSet,
} from '@plone/volto/helpers';

const ImageFromUrl = ({
  image,
  url,
  alt = '',
  className,
  size,
  role,
  ...imageProps
}) => {
  let src = image && image?.scales ? image.download : url;

  if (!src) return null;

  const imageParams = '/@@images/image' + (size ? `/${size}` : '');
  const isInternal = isInternalURL(src);

  src = isInternal ? flattenToAppURL(src) : src;
  if (isInternal && src.indexOf('/@@images') === -1)
    src = `${src}${imageParams}`;

  return (
    <img
      alt={alt}
      role={role ?? 'img'}
      src={src}
      loading="lazy"
      className={className}
      {...(isInternal ? getSrcSet(image?.scales ? image : src, size) : {})}
      {...imageProps}
    />
  );
};

ImageFromUrl.propTypes = {
  image: PropTypes.object,
  url: PropTypes.string,
  alt: PropTypes.string,
  className: PropTypes.string,
  size: PropTypes.string,
  role: PropTypes.string,
};

export default ImageFromUrl;
