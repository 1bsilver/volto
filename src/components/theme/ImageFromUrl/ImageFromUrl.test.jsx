import React from 'react';
import renderer from 'react-test-renderer';

import ImageFromUrl from './ImageFromUrl';

describe('ImageFromUrl', () => {
  it('renders with default values', () => {
    const component = renderer.create(
      <ImageFromUrl url="http://localhost:8080/Plone/photo.png" />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders with props', () => {
    const component = renderer.create(
      <ImageFromUrl
        url="http://localhost:8080/Plone/photo.png"
        alt="Photo"
        className="photo-image"
        size="preview"
        role="presentation"
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
