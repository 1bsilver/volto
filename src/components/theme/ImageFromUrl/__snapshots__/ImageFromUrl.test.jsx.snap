// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageFromUrl renders with default values 1`] = `
<img
  alt=""
  loading="lazy"
  role="img"
  sizes={null}
  src="/photo.png/@@images/image"
  srcSet="/photo.png/@@images/image/large 768w, /photo.png/@@images/image/preview 400w, /photo.png/@@images/image/mini 200w, /photo.png/@@images/image/thumb 128w, /photo.png/@@images/image/tile 64w, /photo.png/@@images/image/icon 32w, /photo.png/@@images/image/listing 16w"
/>
`;

exports[`ImageFromUrl renders with props 1`] = `
<img
  alt="Photo"
  className="photo-image"
  loading="lazy"
  role="presentation"
  sizes="400px"
  src="/photo.png/@@images/image/preview"
  srcSet="/photo.png/@@images/image/preview 400w, /photo.png/@@images/image/mini 200w, /photo.png/@@images/image/thumb 128w, /photo.png/@@images/image/tile 64w, /photo.png/@@images/image/icon 32w, /photo.png/@@images/image/listing 16w"
/>
`;
